# Data Exploration

In [120]:
import pandas as pd
import numpy as np
import re
import clean

## Movies Dataframe
The Movies dataframe contains detailed information about over 940,000 films, including their unique IDs, titles, release dates, poster taglines, short descriptions, durations in minutes, and their ratings.

In [121]:
movies_df = pd.read_csv('../csv/movies.csv')

In [122]:
movies_df.shape

(941597, 7)

This dataset consists of 941,597 rows and 7 columns, providing comprehensive details about a vast collection of films.

In [123]:
movies_df.head()

id                               name    date  \
0  1000001                             Barbie  2023.0   
1  1000002                           Parasite  2019.0   
2  1000003  Everything Everywhere All at Once  2022.0   
3  1000004                         Fight Club  1999.0   
4  1000005                         La La Land  2016.0   

                                            tagline  \
0                  She's everything. He's just Ken.   
1                       Act like you own the place.   
2  The universe is so much bigger than you realize.   
3                           Mischief. Mayhem. Soap.   
4                    Here's to the fools who dream.   

                                         description  minute  rating  
0  Barbie and Ken are having the time of their li...   114.0    3.86  
1  All unemployed, Ki-taek's family takes peculia...   133.0    4.56  
2  An aging Chinese immigrant is swept up in an i...   140.0    4.30  
3  A ticking-time-bomb insomniac and a slippery s...   139.0    4.27  
4  Mia, an aspiring actress, serves lattes to mov...   129.0    4.09

In [124]:
movies_df.dtypes

id               int64
name            object
date           float64
tagline         object
description     object
minute         float64
rating         float64
dtype: object

In [125]:
movies_df['date'] = pd.to_datetime(movies_df['date'], format='%Y', errors='coerce')

The <code>date</code> column is currently stored as a <code>float</code>, which is not suitable for representing release dates; we will convert it to a proper date format for accurate analysis.

In [126]:
movies_df['name'] = movies_df['name'].astype('string')
movies_df['tagline'] = movies_df['tagline'].astype('string')
movies_df['description'] = movies_df['description'].astype('string')

Converting <code>object</code> types to <code>string</code>.

In [127]:
movies_df.dtypes

id                      int64
name           string[python]
date           datetime64[ns]
tagline        string[python]
description    string[python]
minute                float64
rating                float64
dtype: object

Result of the conversions.

### Cleaning text columns

In [128]:
# Rimuove spazi vuoti e converte in minuscolo per le colonne 'name', 'tagline' e 'description'
movies_df['name'] = movies_df['name'].str.strip().str.lower()
movies_df['tagline'] = movies_df['tagline'].str.strip().str.lower()
movies_df['description'] = movies_df['description'].str.strip().str.lower()

In [129]:
columns_to_clean = ['description', 'name', 'tagline']
characters_to_replace = ['.', '-', ',', '#']
patterns_to_replace = [ r'^-+$',          # Una o più linee di trattini
                        r'^-placeholder-$', # La parola "placeholder" esatta
                        r'^-Placeholder-$', # La parola "Placeholder" con iniziale maiuscola
]

movies_df = clean.clean_data(movies_df, columns_to_clean, characters_to_replace, patterns_to_replace)

Clean the <code>movies_df</code> data in the <code>name</code>, <code>tagline</code>, and <code>description</code> columns, removing unnecessary spaces and flattening the format to lowercase. Next, unwanted information, such as specific characters and placeholder words, is removed and replaced with missing values <code>(NaN)</code>.

### Check for <i>null</i> values

In [130]:
movies_df.isna().sum()

id                  0
name               49
date            91913
tagline        802220
description    160852
minute         181570
rating         850598
dtype: int64

In [131]:
movies_df[movies_df['name'].isna()]

id  name       date                                 tagline  \
287514  1287515  <NA> 2015-01-01                                    <NA>   
307905  1307906  <NA> 2018-01-01          they're all gonna laugh at you   
320567  1320568  <NA>        NaT                                    <NA>   
323158  1323159  <NA>        NaT                                    <NA>   
330334  1330335  <NA>        NaT                                    <NA>   
336644  1336645  <NA> 2020-01-01                                    <NA>   
417362  1417363   NaN 2020-01-01                                    <NA>   
424121  1424122   NaN 2011-01-01                                    <NA>   
457879  1457880   NaN 2022-01-01             what is youth but vastness?   
490622  1490623  <NA> 2000-01-01                                    <NA>   
523080  1523081  <NA> 2012-01-01                                    <NA>   
574690  1574691  <NA> 2024-01-01                           100% accurate   
583273  1583274  <NA> 2018-01-01                                    <NA>   
602779  1602780  <NA> 2022-01-01                                    <NA>   
617642  1617643  <NA>        NaT                                    <NA>   
642251  1642252  <NA> 2017-01-01                                    <NA>   
646520  1646521  <NA> 2008-01-01                                    <NA>   
648185  1648186  <NA>        NaT                                    <NA>   
666817  1666818  <NA>        NaT                                    <NA>   
680689  1680690  <NA> 2019-01-01                                    <NA>   
720294  1720295  <NA>        NaT                                    <NA>   
724910  1724911  <NA>        NaT                                    <NA>   
725369  1725370  <NA>        NaT                                    <NA>   
741481  1741482  <NA>        NaT                                    <NA>   
759524  1759525  <NA> 2023-01-01                                    <NA>   
764179  1764180  <NA>        NaT                                    <NA>   
783695  1783696  <NA>        NaT                                    <NA>   
787774  1787775  <NA> 2012-01-01                                    <NA>   
803487  1803488   NaN        NaT                                    <NA>   
826558  1826559  <NA> 2023-01-01                                    <NA>   
840337  1840338  <NA>        NaT                                    <NA>   
859402  1859403   NaN 2024-01-01                         i ran from iran   
859598  1859599  <NA> 2023-01-01                                  leeds!   
865294  1865295  <NA>        NaT                                    <NA>   
865295  1865296  <NA>        NaT                                    <NA>   
873710  1873711  <NA> 2023-01-01                                    <NA>   
881201  1881202  <NA>        NaT                                    <NA>   
883228  1883229  <NA>        NaT                                    <NA>   
885230  1885231  <NA> 2023-01-01                                    <NA>   
892725  1892726  <NA>        NaT  time-travel, bloodline and time-travel   
892726  1892727  <NA>        NaT                                    <NA>   
894771  1894772  <NA>        NaT                                    <NA>   
898056  1898057  <NA>        NaT                                    <NA>   
898057  1898058  <NA>        NaT                                    <NA>   
898058  1898059  <NA>        NaT                                    <NA>   
898059  1898060  <NA>        NaT                                    <NA>   
904117  1904118  <NA>        NaT                                    <NA>   
904118  1904119  <NA>        NaT                                    <NA>   
904119  1904120  <NA>        NaT                                    <NA>   

                                              description  minute  rating  
287514  none is a short film that explores the balance...     4.0     NaN  
307905  a surreal visual film about the inner feelings...   

In [132]:
movies_df = movies_df.dropna(subset=['name'])
#rimuovo righe con name nullo

We chose to remove tuples with <u>**null**</u> <code>name</code>.

### Check for <i>duplicate</i> values

In [133]:
movies_df.duplicated().sum()

np.int64(0)

In [134]:
movies_df['id'].duplicated().any()

np.False_

In [135]:
# Trova i film con nomi duplicati
movies_df['name'].duplicated().any()

np.True_

In [136]:
duplicati = movies_df[movies_df['name'].duplicated(keep=False)]

duplicati = duplicati.sort_values(by='name')

duplicati

id       name       date  \
524259  1524260         #1 2022-01-01   
526951  1526952         #1 2011-01-01   
393181  1393182         #1 2009-01-01   
204330  1204331         #2 1993-01-01   
885075  1885076         #2 2023-01-01   
...         ...        ...        ...   
676051  1676052  정브르의 동물일기 2021-01-01   
848930  1848931   조폭의 브이로그 2023-01-01   
884959  1884960   조폭의 브이로그 2023-01-01   
782298  1782299    코리안 커넥션 1990-01-01   
871369  1871370    코리안 커넥션        NaT   

                                                  tagline  \
524259  a modern story of suspense... in their biggest...   
526951                                               <NA>   
393181                                               <NA>   
204330                                               <NA>   
885075                                               <NA>   
...                                                   ...   
676051                                               <NA>   
848930                                               <NA>   
884959                                               <NA>   
782298                                               <NA>   
871369                                               <NA>   

                                              description  minute  rating  
524259  a broke comic shop owner and an obsessive coll...     8.0     NaN  
526951                     short movie by tomoyasu murata     5.0     NaN  
393181  you're running and you're running away, but yo...     4.0     NaN  
204330  joost rekveld's first film is an hommage to th...    12.0     NaN  
885075                                 subject #2's death     2.0     NaN  
...                                                   ...     ...     ...  
676051                                               <NA>     NaN     NaN  
848930                                               <NA>     NaN     NaN  
884959                                               <NA>    57.0     NaN  
782298                                               <NA>   117.0     NaN  
871369                                               <NA>   117.0     NaN  

[212404 rows x 7 columns]

We found more than 200,000 tuples with duplicate <code>name</code> value, now we will analyze them.

In [137]:
# Rimuovere i duplicati dal DataFrame
movies_df = clean.remove_duplicates(movies_df, clean.get_columns_conditions(movies_df))

We decided to remove duplicates with: 

<ul>
  <li>same <code>name</code>, <code>description</code>, <code>date</code>, <code>tagline</code>, <code>minute</code>, <code>rating</code></li>
  <li>same <code>name</code> but null <code>description</code>, <code>date</code>, <code>tagline</code>, <code>minute</code>, <code>rating</code></li>
  <li>same <code>name</code>, <code>date</code>, <code>description</code> but null <code>tagline</code>, <code>minute</code>, <code>rating</code></li>
  <li>...</li>
</ul>
keeping the first occurrence.

In [138]:
movies_df[(movies_df['rating'] > 5) | (movies_df['rating'] < 0)].shape[0]

0

We checked that in the <code>rating</code> column there are no values ​​outside the range.

---

## Countries Dataframe
This dataset contains over 693,000 entries with the columns 'id', representing the film ID, and 'country', indicating one of the originating countries, with the possibility of multiple countries per film.

In [139]:
countries_df = pd.read_csv('../csv/countries.csv')

In [140]:
countries_df.shape

(693476, 2)

This dataset consists of 693,476 rows and 2 columns, providing detailed information about the countries associated with each film.

In [141]:
countries_df.head()

id      country
0  1000001           UK
1  1000001          USA
2  1000002  South Korea
3  1000003          USA
4  1000004      Germany

In [142]:
countries_df['country'] = countries_df['country'].astype('string')

In [143]:
# Verifica il tipo di dati della colonna 'country'
countries_df.dtypes

id                  int64
country    string[python]
dtype: object

In [144]:
countries_df.isna().sum()

id         0
country    0
dtype: int64

There are no duplicate entries in the dataset.

We also verified that the countries were not duplicated under different names.

---

## Actors Dataframe
This dataset contains over 5.7 million entries with the columns 'id', representing the film ID, 'name', indicating the name of the actor, and 'role', specifying the character played.

In [145]:
actors_df = pd.read_csv('../csv/actors.csv')

In [146]:
actors_df.shape

(5798450, 3)

This dataset consists of 5,798,450 rows and 3 columns, providing detailed information about the actors, their roles, and the films they appear in.

In [147]:
actors_df.head()

id               name    role
0  1000001      Margot Robbie  Barbie
1  1000001       Ryan Gosling     Ken
2  1000001    America Ferrera  Gloria
3  1000001  Ariana Greenblatt   Sasha
4  1000001           Issa Rae  Barbie

In [148]:
actors_df['name'] = actors_df['name'].astype('string')
actors_df['role'] = actors_df['role'].astype('string')

actors_df.dtypes

id               int64
name    string[python]
role    string[python]
dtype: object

In [149]:
actors_df.isna().sum()

id            0
name          4
role    1361559
dtype: int64

In [150]:
actors_df[actors_df['name'].isnull()]

id  name            role
4145738  1443629  <NA>            <NA>
4281100  1469981  <NA>            Self
4306960  1474958  <NA>  Cinematography
5430275  1773264  <NA>            <NA>

In [151]:
actors_df = actors_df.dropna(subset=['name'])

Tuples with <code>name</code> null have been deleted

---

## Crew Dataframe
This dataset contains over 4.7 million entries with the columns 'id', representing the film ID, 'role', indicating the role (e.g., director), and 'name', specifying the name of the person.

In [152]:
crew_df = pd.read_csv('../csv/crew.csv')

In [153]:
crew_df.shape

(4720183, 3)

This dataset consists of 4,720,183 rows and 3 columns, providing detailed information about the crew members, their roles, and the films they are associated with.

In [154]:
crew_df.head()

id      role            name
0  1000001  Director    Greta Gerwig
1  1000001  Producer    Tom Ackerley
2  1000001  Producer   Margot Robbie
3  1000001  Producer  Robbie Brenner
4  1000001  Producer    David Heyman

In [155]:
crew_df['role'] = crew_df['role'].astype('string')
crew_df['name'] = crew_df['name'].astype('string')

crew_df.dtypes

id               int64
role    string[python]
name    string[python]
dtype: object

In [156]:
crew_df.isna().sum()

id      0
role    0
name    1
dtype: int64

In [157]:
crew_df[crew_df['name'].isnull()]

id    role  name
4562126  1859397  Writer  <NA>

A null value was found in the <code>name</code> column, but it will be retained as it may represent a missing or incomplete entry that is important for the context of the dataset.

---

## Posters Dataframe
This dataset contains over 940,000 entries with the columns 'id', representing the film ID, and 'link', providing the link to the original film posters.

In [158]:
posters_df = pd.read_csv('../csv/posters.csv')

In [159]:
posters_df.shape

(941597, 2)

This dataset consists of 941,597 rows and 2 columns, providing links to the original film posters associated with each film.

In [160]:
posters_df

id                                               link
0       1000001  https://a.ltrbxd.com/resized/film-poster/2/7/7...
1       1000002  https://a.ltrbxd.com/resized/film-poster/4/2/6...
2       1000003  https://a.ltrbxd.com/resized/film-poster/4/7/4...
3       1000004  https://a.ltrbxd.com/resized/film-poster/5/1/5...
4       1000005  https://a.ltrbxd.com/resized/film-poster/2/4/0...
...         ...                                                ...
941592  1941593                                                NaN
941593  1941594                                                NaN
941594  1941595  https://a.ltrbxd.com/resized/film-poster/1/1/8...
941595  1941596  https://a.ltrbxd.com/resized/film-poster/1/1/8...
941596  1941597                                                NaN

[941597 rows x 2 columns]

In [161]:
posters_df['link'] = posters_df['link'].astype('string')

posters_df.dtypes

id               int64
link    string[python]
dtype: object

In [162]:
posters_df.isna().sum()

id           0
link    180712
dtype: int64

In [163]:
posters_df = posters_df.dropna(subset=['link'])

Removes all rows from the <code>posters_df</code> that have a <code>NaN</code> value in the link column.

In [164]:
posters_df.duplicated().sum()

np.int64(0)

---

## Releases Dataframe
This dataset contains over 13 million entries with the columns 'id', representing the film ID, 'country', indicating the country of release, 'date', showing the release date, 'type', specifying the release type (e.g., Theatrical, Digital), and 'rating', representing the rating received in that country (e.g., PG, etc.).

In [165]:
releases_df = pd.read_csv('../csv/releases.csv')

In [166]:
releases_df.shape

(1332782, 5)

This dataset consists of 1,332,782 rows and 5 columns, providing detailed information about the film releases, including the country, date, type, and rating in each country.

In [167]:
releases_df.head()

id    country        date        type rating
0  1000001    Andorra  2023-07-21  Theatrical    NaN
1  1000001  Argentina  2023-07-20  Theatrical    ATP
2  1000001  Australia  2023-07-19  Theatrical     PG
3  1000001  Australia  2023-10-01     Digital     PG
4  1000001    Austria  2023-07-20  Theatrical    NaN

In [168]:
releases_df['country'] = releases_df['country'].astype('string')
releases_df['type'] = releases_df['type'].astype('string')
releases_df['rating'] = releases_df['rating'].astype('string')
releases_df['date'] = pd.to_datetime(releases_df['date'])

releases_df.dtypes

id                  int64
country    string[python]
date       datetime64[ns]
type       string[python]
rating     string[python]
dtype: object

The date column is currently stored as an object, which is not suitable for representing release dates; we will convert it to a proper date format for accurate analysis.

In [169]:
releases_df.isna().sum()

id              0
country         0
date            0
type            0
rating     998802
dtype: int64

There are many null values in the 'rating' column, but no action will be taken as they will be left as is, since they may represent cases where a rating was not assigned in certain countries.

In [170]:
releases_df.duplicated().sum()

np.int64(0)

---

## Oscar Awards Dataframe
This dataset contains over 10,000 entries about Oscar candidates, with the columns 'year_film', representing the year the film was issued, 'year_ceremony', indicating the year of the ceremony, 'category', specifying the category (e.g., actor), 'name', showing the name of the person, 'film', listing the title of the film, and 'winner', indicating whether they won or not.

In [171]:
oscar_awards_df = pd.read_csv('../csv/the_oscar_awards.csv')

In [172]:
oscar_awards_df.shape

(10889, 7)

This dataset consists of 10,889 rows and 7 columns, providing detailed information about Oscar candidates, including the year of the film, ceremony, category, person, film title, and whether the candidate won.

In [173]:
oscar_awards_df.head()

year_film  year_ceremony  ceremony category                 name  \
0       1927           1928         1    ACTOR  Richard Barthelmess   
1       1927           1928         1    ACTOR        Emil Jannings   
2       1927           1928         1  ACTRESS       Louise Dresser   
3       1927           1928         1  ACTRESS         Janet Gaynor   
4       1927           1928         1  ACTRESS       Gloria Swanson   

               film  winner  
0         The Noose   False  
1  The Last Command    True  
2   A Ship Comes In   False  
3        7th Heaven    True  
4    Sadie Thompson   False

In [174]:
oscar_awards_df['category'] = oscar_awards_df['category'].astype('string')
oscar_awards_df['name'] = oscar_awards_df['name'].astype('string')
oscar_awards_df['film'] = oscar_awards_df['film'].astype('string')

oscar_awards_df.dtypes

year_film                 int64
year_ceremony             int64
ceremony                  int64
category         string[python]
name             string[python]
film             string[python]
winner                     bool
dtype: object

In [175]:
oscar_awards_df.isna().sum()

year_film          0
year_ceremony      0
ceremony           0
category           0
name               5
film             319
winner             0
dtype: int64

In [176]:
oscar_awards_df[oscar_awards_df['name'].isnull()]

year_film  year_ceremony  ceremony                          category  \
10513       2020           2021        93  JEAN HERSHOLT HUMANITARIAN AWARD   
10514       2020           2021        93  JEAN HERSHOLT HUMANITARIAN AWARD   
10635       2021           2022        94  JEAN HERSHOLT HUMANITARIAN AWARD   
10759       2022           2023        95  JEAN HERSHOLT HUMANITARIAN AWARD   
10885       2023           2024        96  JEAN HERSHOLT HUMANITARIAN AWARD   

       name  film  winner  
10513  <NA>  <NA>    True  
10514  <NA>  <NA>    True  
10635  <NA>  <NA>    True  
10759  <NA>  <NA>    True  
10885  <NA>  <NA>    True

In [177]:
oscar_awards_df[oscar_awards_df['film'].isnull()]

year_film  year_ceremony  ceremony                          category  \
16          1927           1928         1               ENGINEERING EFFECTS   
18          1927           1928         1               ENGINEERING EFFECTS   
31          1927           1928         1           WRITING (Title Writing)   
32          1927           1928         1           WRITING (Title Writing)   
33          1927           1928         1                     SPECIAL AWARD   
...          ...            ...       ...                               ...   
10763       2022           2023        95            GORDON E. SAWYER AWARD   
10885       2023           2024        96  JEAN HERSHOLT HUMANITARIAN AWARD   
10886       2023           2024        96                    HONORARY AWARD   
10887       2023           2024        96                    HONORARY AWARD   
10888       2023           2024        96                    HONORARY AWARD   

                                                    name  film  winner  
16                                        Ralph Hammeras  <NA>   False  
18                                      Nugent Slaughter  <NA>   False  
31                                        Joseph Farnham  <NA>    True  
32                                    George Marion, Jr.  <NA>   False  
33                                          Warner Bros.  <NA>    True  
...                                                  ...   ...     ...  
10763                                          Iain Neil  <NA>    True  
10885                                               <NA>  <NA>    True  
10886  To Angela Bassett, who has inspired audiences ...  <NA>    True  
10887  To Mel Brooks, for his comedic brilliance, pro...  <NA>    True  
10888  To Carol Littleton, whose commitment to her cr...  <NA>    True  

[319 rows x 7 columns]

Null values were found, but will be kept as it may represent an important entry for the context of the dataset.

In [178]:
invalid_oscar = oscar_awards_df[oscar_awards_df['name'].isna() & oscar_awards_df['film'].isna()]

invalid_oscar

year_film  year_ceremony  ceremony                          category  \
10513       2020           2021        93  JEAN HERSHOLT HUMANITARIAN AWARD   
10514       2020           2021        93  JEAN HERSHOLT HUMANITARIAN AWARD   
10635       2021           2022        94  JEAN HERSHOLT HUMANITARIAN AWARD   
10759       2022           2023        95  JEAN HERSHOLT HUMANITARIAN AWARD   
10885       2023           2024        96  JEAN HERSHOLT HUMANITARIAN AWARD   

       name  film  winner  
10513  <NA>  <NA>    True  
10514  <NA>  <NA>    True  
10635  <NA>  <NA>    True  
10759  <NA>  <NA>    True  
10885  <NA>  <NA>    True

In [179]:
oscar_awards_df = oscar_awards_df.drop(invalid_oscar.index)

The <i>Jean Hersholt Humanitarian Award</i> is a special category for exceptional contributions to humanitarian causes.

Since tuples have no value for <code>name</code>, we decided not to keep them.

---

## Rotten Tomatoes Reviews Dataframe
This dataset contains over 1.1 million reviews from Rotten Tomatoes, with the columns <code>rotten_tomatoes_link</code>, providing the URL of the Rotten Tomatoes page, <code>movie_title</code>, indicating the title of the movie, <code>critic_name</code>, showing the name of the critic, <code>top_critic</code>, indicating if the critic is a top critic, <code>publisher_name</code>, specifying where the review was published, <code>review_type</code>, categorizing the type of review, <code>review_score</code>, representing the score given in the review, <code>review_date</code>, showing the date of the review, and <code>review_content</code>, containing the text of the actual review.

In [180]:
rotten_tomatoes_df = pd.read_csv('../csv/rotten_tomatoes_reviews.csv')

In [181]:
rotten_tomatoes_df.shape

(1129887, 9)

This dataset consists of 1,129,887 rows and 9 columns, providing detailed information about Rotten Tomatoes reviews, including the review URL, movie title, critic details, review score, date, and content.

In [182]:
rotten_tomatoes_df

rotten_tomatoes_link  \
0                  m/0814255   
1                  m/0814255   
2                  m/0814255   
3                  m/0814255   
4                  m/0814255   
...                      ...   
1129882          m/zulu_dawn   
1129883          m/zulu_dawn   
1129884          m/zulu_dawn   
1129885          m/zulu_dawn   
1129886          m/zulu_dawn   

                                               movie_title        critic_name  \
0        Percy Jackson & the Olympians: The Lightning T...    Andrew L. Urban   
1        Percy Jackson & the Olympians: The Lightning T...      Louise Keller   
2        Percy Jackson & the Olympians: The Lightning T...                NaN   
3        Percy Jackson & the Olympians: The Lightning T...       Ben McEachen   
4        Percy Jackson & the Olympians: The Lightning T...        Ethan Alter   
...                                                    ...                ...   
1129882                                          Zulu Dawn      Chuck O'Leary   
1129883                                          Zulu Dawn          Ken Hanke   
1129884                                          Zulu Dawn    Dennis Schwartz   
1129885                                          Zulu Dawn  Christopher Lloyd   
1129886                                          Zulu Dawn     Brent McKnight   

         top_critic                   publisher_name review_type review_score  \
0             False                   Urban Cinefile       Fresh          NaN   
1             False                   Urban Cinefile       Fresh          NaN   
2             False              FILMINK (Australia)       Fresh          NaN   
3             False          Sunday Mail (Australia)       Fresh        3.5/5   
4              True               Hollywood Reporter      Rotten          NaN   
...             ...                              ...         ...          ...   
1129882       False                 Fantastica Daily      Rotten          2/5   
1129883       False  Mountain Xpress (Asheville, NC)       Fresh        3.5/5   
1129884       False    Dennis Schwartz Movie Reviews       Fresh           B+   
1129885       False          Sarasota Herald-Tribune      Rotten        3.5/5   
1129886       False             The Last Thing I See      Rotten            C   

        review_date                                     review_content  
0        2010-02-06  A fantasy adventure that fuses Greek mythology...  
1        2010-02-06  Uma Thurman as Medusa, the gorgon with a coiff...  
2        2010-02-09  With a top-notch cast and dazzling special eff...  
3        2010-02-09  Whether audiences will get behind The Lightnin...  
4        2010-02-10  What's really lacking in The Lightning Thief i...  
...             ...                                                ...  
1129882  2005-11-02                                                NaN  
1129883  2007-03-07  Seen today, it's not only a startling indictme...  
1129884  2010-09-16  A rousing visual spectacle that's a prequel of...  
1129885  2011-02-28  A simple two-act story: Prelude to war, and th...  
1129886  2020-07-09  Rides the line between being a pure artifact o...  

[1129887 rows x 9 columns]

In [183]:
rotten_tomatoes_df.dtypes

rotten_tomatoes_link    object
movie_title             object
critic_name             object
top_critic                bool
publisher_name          object
review_type             object
review_score            object
review_date             object
review_content          object
dtype: object

Many columns are currently stored as <code>objects</code>, we will convert them to more useful formats for manipulation.

In [184]:
rotten_tomatoes_df['review_date'] = pd.to_datetime(rotten_tomatoes_df['review_date'])
rotten_tomatoes_df['movie_title'] = rotten_tomatoes_df['movie_title'].astype('string')
rotten_tomatoes_df['critic_name'] = rotten_tomatoes_df['critic_name'].astype('string')
rotten_tomatoes_df['publisher_name'] = rotten_tomatoes_df['publisher_name'].astype('string')
rotten_tomatoes_df['review_type'] = rotten_tomatoes_df['review_type'].astype('string')
rotten_tomatoes_df['review_score'] = rotten_tomatoes_df['review_score'].astype('string')
rotten_tomatoes_df['review_content'] = rotten_tomatoes_df['review_content'].astype('string')

rotten_tomatoes_df.dtypes

rotten_tomatoes_link            object
movie_title             string[python]
critic_name             string[python]
top_critic                        bool
publisher_name          string[python]
review_type             string[python]
review_score            string[python]
review_date             datetime64[ns]
review_content          string[python]
dtype: object

In [185]:
rotten_tomatoes_df.isna().sum()

rotten_tomatoes_link         0
movie_title                  0
critic_name              18521
top_critic                   0
publisher_name               0
review_type                  0
review_score            305902
review_date                  0
review_content           65778
dtype: int64

In [186]:
rotten_tomatoes_df[rotten_tomatoes_df['review_score'].isnull() &
                           rotten_tomatoes_df['review_content'].isnull() &
                           rotten_tomatoes_df['critic_name'].isnull()]

rotten_tomatoes_link                     movie_title  \
118910           m/alice_sweet_alice  Alice, Sweet Alice (Communion)   
219988              m/callas_forever                  Callas Forever   
323790    m/escape_to_witch_mountain        Escape to Witch Mountain   
349208                    m/flawless                        Flawless   
493769                m/kings_ransom                   King's Ransom   
619413                 m/nurse_betty                     Nurse Betty   
751877  m/saddest_music_in_the_world  The Saddest Music in the World   
751879  m/saddest_music_in_the_world  The Saddest Music in the World   

       critic_name  top_critic                publisher_name review_type  \
118910        <NA>       False                         Film4      Rotten   
219988        <NA>        True    Denver Rocky Mountain News      Rotten   
323790        <NA>        True                Chicago Reader       Fresh   
349208        <NA>       False                     E! Online      Rotten   
493769        <NA>       False                 Hollywood.com       Fresh   
619413        <NA>        True  Atlanta Journal-Constitution       Fresh   
751877        <NA>       False             Premiere Magazine      Rotten   
751879        <NA>       False             Premiere Magazine      Rotten   

       review_score review_date review_content  
118910         <NA>  2003-05-24           <NA>  
219988         <NA>  2004-12-17           <NA>  
323790         <NA>  2000-01-01           <NA>  
349208         <NA>  2000-01-01           <NA>  
493769         <NA>  2005-04-23           <NA>  
619413         <NA>  2000-01-01           <NA>  
751877         <NA>  2004-07-03           <NA>  
751879         <NA>  2004-07-03           <NA>

Reviews without <code>review_score</code>, <code>review_content</code>, <code>critic_name</code> were found but will be kept because they may be relevant in the future.

# Final thoughts on dataframes

In [187]:
# Trova tutti i titoli di film unici nel dataset di Rotten Tomatoes
unique_rotten_tomatoes_movies = rotten_tomatoes_df['movie_title'].unique()

# Trova i titoli di film unici nel dataset dei film
unique_movies = movies_df['name'].unique()

# Converti tutti i titoli di film in stringhe e poi in minuscolo per il confronto
unique_rotten_tomatoes_movies = [str(title).lower() for title in unique_rotten_tomatoes_movies]
unique_movies = [str(title).lower() for title in unique_movies]

# Trova i film che sono presenti in entrambi i dataset
common_movies = set(unique_rotten_tomatoes_movies).intersection(set(unique_movies))

# Stampa il numero di film trovati
print(f"Number of unique films on Rotten Tomatoes: {len(unique_rotten_tomatoes_movies)}")
print(f"Number of unique movies in the movie dataset: {len(unique_movies)}")
print(f"Number of common films: {len(common_movies)}")

Number of unique films on Rotten Tomatoes: 17100
Number of unique movies in the movie dataset: 786963
Number of common films: 15056


# Saving modified dataframes

In [188]:
movies_df.to_csv('movie_df.csv', index=False)

In [189]:
actors_df.to_csv('actors_df.csv', index=False)

In [190]:
posters_df.to_csv('posters_df.csv', index=False)

In [191]:
oscar_awards_df.to_csv('oscar_awards_df.csv', index=False)